src/cachemachine/rubinrepoman.py

```diff
@@ -65,124 +66,132 @@ async def desired_images(
+        #  was filtered out in _allow_image (which is nothing for this class,
+        #  but might be overridden in a subclass).
+
+        # Now we're going to sort each of the daily/weekly/release lists,
+        #  and take the requested number of them.  For each of those we will
+        #  get the hash (this is why we're processing the list twice: that's
+        #  an expensive operation, and we only want to bother getting the
+        #  digest for the prepulled images).  Sorting reverse=True will put
+        #  the most recent ones (with the highest semvers) on top.
+
+        for k, v in {
+            "release": self.num_releases,
+            "weekly": self.num_weeklies,
+            "daily": self.num_dailies,
+        }.items():
+            inputlist = sorted(imgmap[k][:v], reverse=True)
+            for img in inputlist:
+                image_hash = await self.docker_client.get_image_hash(img.tag)
+                h_image = RubinImageInfo.from_reference(
+                    img.reference,
+                    recommended_tag=self.recommended_tag,
+                    digest=image_hash,
+                )
+                pull_images.append(self._docker_image_from_image_info(h_image))
         logger.info(f"Returning {pull_images}")
         return DesiredImageList(pull_images, all_images)
 
-    def _friendly_name(self, tag: str) -> str:
-        """Generate the friendly name of an image based on its tag.
-
-        Parameters
-        ----------
-        tag: tag to generate the friendly name of.  Only works on
-          release, weekly, and daily tags.
-
-        Returns
-        -------
-        The friendly name of the tag.
-        """
-        tag_parts = tag.split("_")
-
-        if tag.startswith("d_"):
-            return f"Daily {tag_parts[1]}_{tag_parts[2]}_{tag_parts[3]}"
-        elif tag.startswith("w_"):
-            return f"Weekly {tag_parts[1]}_{tag_parts[2]}"
-        elif tag.startswith("r"):
-            return "Release " + ".".join(tag_parts)
-        else:
-            # Should never reach here...
-            raise Exception(f"Unexpected tag name {tag}")
-
-    def _is_real_release(self, tag: str) -> bool:
-        """We want to be able to reject release RC versions
-        (e.g. r_22_0_0_rc1).
-
-        Our heuristic here is that if the last tag component contains only
-        digits then it is a release version; otherwise it is not.
-        """
-        tag_parts = tag.split("_")
-        if tag.startswith("r") and tag_parts[-1].isdigit():
-            return True
-        return False
+    def _docker_image_from_image_info(
+        self, img: RubinImageInfo, override_name: Optional[str] = None
+    ) -> DockerImage:
+        name = img.display_name
+        if override_name:
+            name = override_name
+        return DockerImage(
+            image_url=img.reference, image_hash=img.digest, name=name
+        )
+
+    def _allow_image(self, img: RubinImageInfo) -> bool:
```

Comment:
Using this subclassing just to exclude some different cycle things seems... weird.  We have all this logic for comparing, why not just put it in there instead of in this subclass thing?

Reply:
Because I understood you to ask for a subclass when I said initially "hey, we need to add the ability for T&S to filter by cycle."